### PR TITLE
Cooldown changes

### LINKED ISSUE #1142: Protocol Update 7 Test Plan

### PR DESCRIPTION
**Pull request by td202**
## Purpose

Closes #1142 #1145

Implement cooldown changes in the node.

## Changes

- Split `bsoConfigureBaker` into `bsoAddValidator` and `bsoUpdateValidator`. Rewrite to factor out the checks and to make the flow of account updates more apparent.
  - The behaviour of `bsoUpdateValidator` is changed in the following case: when the validator's keys are changed, and the new aggregation key duplicates the aggregation key of some other validator, the key that is actually duplicated is now returned (whereas previously the old key on the account was returned) in the error. This causes a difference in the RejectReason which is reported for such a transaction, and could cause incompatibility if such a difference was observed in P1-P4 (since P5, reject reasons are not included in the block hashing scheme). (This does not seem to have occurred at least on mainnet.)
- Split `bsoConfigureDelegation` into `bsoAddDelegator` and `bsoUpdateDelegator`. Rewrite to factor out the checks and to make the flow of account updates more apparent.
- Extensive tests. To assure backwards compatibility (except in the above noted difference), the tests were also run against the old `bsoConfigureBaker` and `bsoConfigureDelegation` (see the branch: https://github.com/Concordium/concordium-node/tree/backport-tests).

## Checklist

- [ ] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: concordium-consensus/src/Concordium/GlobalState/BakerInfo.hs [hunk 1]
```diff
@@ -174,32 +183,83 @@ data BakerAddResult
...
-          bcuKeys :: !(Maybe BakerKeyUpdate),
-          bcuCapital :: !(Maybe Amount),
-          bcuRestakeEarnings :: !(Maybe Bool),
-          bcuOpenForDelegation :: !(Maybe OpenStatus),
-          bcuMetadataURL :: !(Maybe UrlText),
-          bcuTransactionFeeCommission :: !(Maybe AmountFraction),
-          bcuBakingRewardCommission :: !(Maybe AmountFraction),
-          bcuFinalizationRewardCommission :: !(Maybe AmountFraction)
+-- | Result of remove baker.
+data BakerRemoveResult
```
Comment (drsk0) [inline concordium-consensus/src/Concordium/GlobalState/BakerInfo.hs:187]: How are these changes related to cooldowns?

File: concordium-consensus/src/Concordium/Scheduler.hs [hunk 2]
```diff
@@ -2077,280 +2077,292 @@ handleConfigureBaker
         withDeposit wtc tickGetArgAndBalance chargeAndExecute
       where
         senderAccount = wtc ^. wtcSenderAccount
+        senderAccountIndex = fst senderAccount
+        bid = BakerId senderAccountIndex
         senderAddress = wtc ^. wtcSenderAddress
-        configureAddBakerArg =
```
Comment (drsk0) [inline concordium-consensus/src/Concordium/Scheduler.hs:2081]: Why do we have almost complete rewrites of these functions induced by the cooldown changes?

**Review by drsk0 — DISMISSED** [review]
Imho, this PR is way to big. There is no way I can give you a serious review  over a diff of more than 7000 lines of code. Could you please break it down in thematic grouped small PR's. For example, we could introduce

- The new boilerplate that is needed and tests for those changes
- For each new data structure representing cooldown queues we can have one PR together with its unit tests
- The changes in accounts + tests
- The changes in block state + tests
- The changes in the scheduler code  + tests
- etc...
I'm aware that this leads to unimplemented code paths in main, but these should be well guarded behind the protocol update to version 7.

File: concordium-consensus/src/Concordium/GlobalState/BlockState.hs [hunk 3]
```diff
@@ -1396,6 +1503,18 @@ class (BlockStateQuery m) => BlockStateOperations m where
     -- | Get whether a protocol update is effective
     bsoIsProtocolUpdateEffective :: UpdatableBlockState m -> m Bool
 
+    -- | A snapshot of the block state that can be used to roll back to a previous state.
+    type StateSnapshot m
```
[the remark this answers is not in the document]
Comment (td202) [reply concordium-consensus/src/Concordium/GlobalState/BlockState.hs:1507]: I think probably it would lead to a dependency issue. In any case, I prefer the type to be abstract.

File: concordium-consensus/src/Concordium/Scheduler.hs [hunk 4]
```diff
@@ -2020,29 +2020,32 @@ handleAddBaker wtc abElectionVerifyKey abSignatureVerifyKey abAggregationVerifyK
...
-    = ConfigureAddBakerCont
-        { cbcCapital :: !Amount,
-          cbcRestakeEarnings :: !Bool,
-          cbcOpenForDelegation :: !OpenStatus,
-          cbcKeysWithProofs :: !BakerKeysWithProofs,
-          cbcMetadataURL :: !UrlText,
-          cbcTransactionFeeCommission :: !AmountFraction,
-          cbcBakingRewardCommission :: !AmountFraction,
-          cbcFinalizationRewardCommission :: !AmountFraction
+data ConfigureBakerCont (av :: AccountVersion)
```
[the remark this answers is not in the document]
Comment (td202) [reply concordium-consensus/src/Concordium/Scheduler.hs:2023]: Values should be short-lived, and at some point there was more of a reason, but now there's not much computation involved in these values so making them strict should be fine.

File: concordium-consensus/src/Concordium/Scheduler.hs [hunk 5]
```diff
@@ -2077,280 +2080,292 @@ handleConfigureBaker
...
+                CTrue True -> do
+                    -- Remove the validator if necessary.
+                    -- We know this will succeed because:
+                    -- 1. The account is a validator.
+                    -- 2. Flexible cooldown is enabled, meaning:
+                    --    * The account cannot have a pending change (which would cause failure)
+                    --    * Setting the capital to 0 will remove the validator immediately,
+                    --      moving the old staked capital to pre-pre-cooldown.
+                    cm <- getChainMetadata
+                    _ <- updateValidator (slotTime cm) senderAccountIndex BI.validatorRemove
```
[the remark this answers is not in the document]
Comment (td202) [reply concordium-consensus/src/Concordium/Scheduler.hs]: Done.

File: concordium-consensus/src/Concordium/Scheduler.hs [hunk 6]
```diff
@@ -2077,280 +2080,292 @@ handleConfigureBaker
...
+                    CTrue True -> do
+                        -- Remove the delegator if necessary.
+                        -- We know this will succeed because:
+                        -- 1. The account is a delegator.
+                        -- 2. Flexible cooldown is enabled, meaning:
+                        --    * The account cannot have a pending change (which would cause failure)
+                        --    * Setting the capital to 0 will remove the delegator immediately,
+                        --      moving the old staked capital to pre-pre-cooldown.
+                        cm <- getChainMetadata
+                        _ <- updateDelegator (slotTime cm) senderAccountIndex BI.delegatorRemove
```
[the remark this answers is not in the document]
Comment (td202) [reply concordium-consensus/src/Concordium/Scheduler.hs]: Done.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 7]
```diff
@@ -1747,220 +2035,496 @@ doConstrainBakerCommission pbs ai ranges = do
...
+        m' (BlockStatePointers pv)
+    addToPrePreCooldowns bsp = do
+        -- Add the account to the pre-pre-cooldowns list.
+        newAccountsInCooldown <-
+            (accountsInCooldown . prePreCooldown)
+                (consAccountList ai)
+                (bspAccountsInCooldown bsp)
+        return bsp{bspAccountsInCooldown = newAccountsInCooldown}
+
+data CooldownRemovals = CooldownRemovals
```
Comment (drsk0) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:2483]: Should these fields be strict?

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 8]
```diff
@@ -1747,220 +2035,496 @@ doConstrainBakerCommission pbs ai ranges = do
...
+      -- | Whether the pre cooldown was removed.
+      crPreCooldown :: Bool,
+      -- | If all cooldowns were removed, this is the previous timestamp of the earliest cooldown.
+      crCooldown :: Maybe Timestamp
+    }
+
+-- | Determine if a change in cooldowns requires global updates to the indexes.
+--  The change should arise from (possibly) reactivating stake from cooldown.
+cooldownRemovals ::
+    Maybe CooldownQueue.Cooldowns -> Maybe CooldownQueue.Cooldowns -> CooldownRemovals
```
Comment (drsk0) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:2495]: Could you comment the inputs? What's the first/second cooldown queue?

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 9]
```diff
@@ -2033,7 +2597,7 @@ doUpdateBakerStake pbs ai newStake = do
                             storePBS pbs bsp{bspAccounts = newAccounts}
                     case compare newStake sdStakedCapital of
                         LT ->
-                            if newStake < bakerStakeThreshold
+                            if newStake < max 1 bakerStakeThreshold
```
Comment (drsk0) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:2600]: Was this a bug or is the change induced by the cooldown changes?

**Review by Victor-N-Suadicani — COMMENTED** [review]
I have not had time to look at the whole thing either, here are just a few comments

**Review by limemloh — APPROVED** [review]

File: concordium-consensus/tests/scheduler/SchedulerTests/Delegation.hs [hunk 10]
```diff
@@ -527,6 +572,348 @@ testCase9 _ pvString =
...
+            assertEqual
+                "Baker account should not have changed except nonce and balance"
+                ( initialBaker4
+                    & Transient.accountNonce .~ 2
+                    & Transient.accountAmount -~ Helpers.energyToAmount (Cost.configureDelegationCost + Cost.baseCost 81 1)
+                )
+                updatedBaker4
+
+-- | Reduce stake while in cooldown.
+testCase12 ::
```
Comment (drsk0) [inline concordium-consensus/tests/scheduler/SchedulerTests/Delegation.hs:756]: Can we have more descriptive names for these tests?

**Comment by Victor-N-Suadicani** [comment]
I pushed [a fix to the consensus](https://github.com/Concordium/concordium-node/pull/1175/commits/ddf852e9d0e9805214b71dcaae7412ba50c899be), because it didn't actually recognize the P7 protocol update. I'm not really sure why I didn't see this when I was testing the P6 to P7 update, where it seemed to work fine, but for P7 to P7 it did not work. In any case, it should be fixed now. I also changed it from using a wildcard pattern match to avoid future problems like this.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 11]
```diff
@@ -1480,245 +1491,522 @@ redelegatePassive accounts (DelegatorId accId) =
...
+        return $ bspBirkParameters bsp & birkActiveBakers .~ newPABref
+    let poolInfo =
+            BaseAccounts.BakerPoolInfo
+                { _poolOpenStatus = vaOpenForDelegation,
+                  _poolMetadataUrl = vaMetadataURL,
+                  _poolCommissionRates = vaCommissionRates
+                }
+    let bakerInfo = bakerKeyUpdateToInfo bid vaKeys
+    let bakerInfoEx = BaseAccounts.BakerInfoExV1 bakerInfo poolInfo
+    acc <- fromJust <$> Accounts.indexedAccount ai (bspAccounts bsp)
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:1615]: I believe it shouldn't happen due to the first precondition

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 12]
```diff
@@ -1480,245 +1491,522 @@ redelegatePassive accounts (DelegatorId accId) =
...
+    -- | Current block timestamp
+    Timestamp ->
+    AccountIndex ->
+    ValidatorUpdate ->
+    MTL.ExceptT ValidatorConfigureFailure m ([BakerConfigureUpdateChange], PersistentBlockState (MPV m))
+newUpdateValidator pbs curTimestamp ai vu@ValidatorUpdate{..} = do
+    bsp <- loadPBS pbs
+    -- Cannot fail: account must exist.
+    acc <- fromJust <$> Accounts.indexedAccount ai (bspAccounts bsp)
+    -- Cannot fail: account must be a registered baker.
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:1819]: Same as above, due to the first precondition

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs (hunk 11 above)
[the remark this answers is not in the document]
Comment (drsk0) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:1615]: But where's the first precondition enforced? I think it would be nice in these cases to point to the code to 'proof' that the error case can't happen.

[the remark this answers is not in the document]
Comment (drsk0) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:1615]: Maybe worth pointing this out in a comment. I think it's generally good practice to always add a comment on the safety when using non-total functions.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 13]
```diff
@@ -1480,245 +1491,522 @@ redelegatePassive accounts (DelegatorId accId) =
...
-            Just acc ->
-                accountBaker acc >>= \case
-                    Nothing -> MTL.throwError BCInvalidBaker
-                    Just bkr -> return bkr
-    modifyAccount' updAcc = do
-        bsp <- MTL.get
-        newAccounts <- liftBSO $ Accounts.updateAccountsAtIndex' updAcc ai (bspAccounts bsp)
-        MTL.put bsp{bspAccounts = newAccounts}
-    ifPresent Nothing _ = return return
+    ifPresent Nothing _ = return
```
[the remark this answers is not in the document]
Comment (drsk0) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:1833]: You're right, would maybe also be worth adding a comment, explaining that this is the trivial update and not failing.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs (hunk 9 above)
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:2600]: It was a bug.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 14]
```diff
@@ -1480,245 +1491,522 @@ redelegatePassive accounts (DelegatorId accId) =
...
+--  6. If the baking reward commission is supplied:
+--
+--        (1) if the commission does not fall within the current range according to the chain
+--            parameters, return @VCFBakingRewardCommissionNotInRange@; otherwise,
+--
+--        (2) update the account's baking reward commission rate to the the supplied value @brc@;
+--
+--        (3) append @BakerConfigureBakingRewardCommission brc@ to @events@.
+--
+--  6. If the finalization reward commission is supplied:
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs]: Fixed.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs (hunk 13 above)
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:1835]: Added comment explaining `ifPresent`.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 15]
```diff
@@ -1747,220 +2035,496 @@ doConstrainBakerCommission pbs ai ranges = do
...
-        Left e -> return (e, pbs)
-        Right newBirkParams -> (DCSuccess [] did,) <$> storePBS pbs newBirkParams
+    DelegatorAdd ->
+    MTL.ExceptT DelegatorConfigureFailure m (PersistentBlockState (MPV m))
+newAddDelegator pbs ai da@DelegatorAdd{..} = do
+    bsp <- loadPBS pbs
+    addDelegatorChecks bsp da
+    newBirkParameters <- do
+        pab <- refLoad $ bspBirkParameters bsp ^. birkActiveBakers
+        -- Cannot fail: the delegation target is valid.
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs]: Done.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 16]
```diff
@@ -1747,220 +2035,496 @@ doConstrainBakerCommission pbs ai ranges = do
...
+updateDelegatorChecks bsp oldDelegator DelegatorUpdate{..} = do
+    -- Check that the delegation target is valid and open.
+    -- This returns @Just (baker, sameBaker)@ if the (new) delegation target is a baker (@baker@),
+    -- with @sameBaker@ indicating whether the delegator is still delegating to the same pool.
+    -- If the target is passive delegation, it returns @Nothing@.
+    mTargetBaker <- case duDelegationTarget of
+        Nothing -> case oldDelegator ^. BaseAccounts.delegationTarget of
+            Transactions.DelegatePassive -> return Nothing
+            Transactions.DelegateToBaker (BakerId baid) -> do
+                -- Cannot fail: the account must delegate to a valid baker.
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs]: Done.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 17]
```diff
@@ -1747,220 +2035,496 @@ doConstrainBakerCommission pbs ai ranges = do
...
+--       (2) Append @DelegationConfigureRestakeEarnings restakeEarnings@ to @events@.
+--
+--  3. If the delegated capital is specified as @capital@: if there is a pending change to the
+--     delegator's stake, return 'DCFChangePending'; otherwise:
+--
+--       * If the new capital is 0, mark the delegator as pending removal at the slot timestamp
+--         plus the delegator cooldown chain parameter, and append
+--         @DelegationConfigureStakeReduced capital@ to @events@; otherwise
+--
+--       * If the the new capital is less than the current staked capital (but not 0), mark the
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs]: Fixed.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 18]
```diff
@@ -1747,220 +2035,496 @@ doConstrainBakerCommission pbs ai ranges = do
...
+--       (2) If the delegation target is baker id @bid@, but the baker does not exist, return
+--           @DCFInvalidDelegationTarget bid@.
+--
+--       (3) Update the active bakers index to: remove the delegator and delegated amount from the
+--           old baker pool, and add the delegator and delegated amount to the new baker pool.
+--           (Note, the total delegated amount is unchanged at this point.)
+--
+--       (4) Update the account to record the new delegation target.
+--
+--       (5) Append @DelegationConfigureDelegationTarget target@ to @events@. [N.B. if the target is
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs]: Fixed.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 19]
```diff
@@ -1747,220 +2035,496 @@ doConstrainBakerCommission pbs ai ranges = do
...
-                -- Update the account with the new delegation target.
-                modifyAccount (setAccountDelegationTarget target)
-            MTL.tell [DelegationConfigureDelegationTarget target]
-        return oldTarget
-    updateRestakeEarnings = forM_ dcuRestakeEarnings $ \restakeEarnings -> do
-        acctDlg <- getAccountOrFail
-        unless (acctDlg ^. BaseAccounts.delegationStakeEarnings == restakeEarnings) $ do
-            modifyAccount (setAccountRestakeEarnings restakeEarnings)
+    flexibleCooldown = sSupportsFlexibleCooldown (accountVersion @(AccountVersionFor pv))
+    ifPresent Nothing _ = return
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:2375]: Done.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 20]
```diff
@@ -1747,220 +2035,496 @@ doConstrainBakerCommission pbs ai ranges = do
...
+                        newCooldowns <- applyCooldownRemovalsGlobally ai removals (bspAccountsInCooldown bsp1)
+                        return (bsp1{bspAccountsInCooldown = newCooldowns}, acc1)
+    onActiveBakers bsp f = do
+        newPABRef <- refMake =<< f =<< refLoad (bspBirkParameters bsp ^. birkActiveBakers)
+        return bsp{bspBirkParameters = bspBirkParameters bsp & birkActiveBakers .~ newPABRef}
+    addToPrePreCooldowns ::
+        (MonadBlobStore m', PVSupportsFlexibleCooldown pv) =>
+        BlockStatePointers pv ->
+        m' (BlockStatePointers pv)
+    addToPrePreCooldowns bsp = do
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:2478]: I think it's not that necessary in this case, because it's only a few lines of code.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs (hunk 7 above)
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:2488]: Yes. Made them strict.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs (hunk 8 above)
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:2502]: Done.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/Bakers.hs [hunk 21]
```diff
@@ -497,6 +518,24 @@ addDelegator (DelegateToBaker bid) did amt pab =
             newActiveBakers <- Trie.insert bid pad' (pab ^. activeBakers)
             return $ Right $ pab & activeBakers .~ newActiveBakers
 
+-- | Add a delegator to the persistent active bakers at a particular target.
+--  It is assumed that the delegator is not already delegated to this target.
+--  If the target is a baker, then the baker MUST be in the active bakers.
+--
+--  IMPORTANT: This does not update the total active capital!
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/Bakers.hs:525]: Because when a delegator or baker is added, the total active stake increases. One should therefore be aware that the total active stake is not updated by this function.

File: concordium-consensus/tests/scheduler/SchedulerTests/Delegation.hs (hunk 10 above)
Comment (eb-concordium) [reply concordium-consensus/tests/scheduler/SchedulerTests/Delegation.hs]: Good idea. Added descriptive names.

File: CHANGELOG.md [hunk 22]
```diff
@@ -20,6 +20,16 @@
...
+- Changes to stake cooldown behavior in protocol version 7:
+  - When stake is reduced or removed from a validator or delegator, it becomes
+    inactive, and is not counted for future stake calculations. The inactive
+    stake is not spendable, but is released after a cooldown period elapses.
+  - Changes to validators and delegators can be made while stake is in cooldown,
+    including changing the stake, or changing directly between validator and
+    delegator.
+- Fix a bug where a configure-validator transaction that is rejected for having
+  a duplicate aggregation key would report the old key for the validator,
+  rather than the key that is a duplicate.
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply CHANGELOG.md:39]: I believe the answer is yes. The changelog shouldn't contain implementation details.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/Bakers.hs [hunk 23]
```diff
@@ -497,6 +518,24 @@ addDelegator (DelegateToBaker bid) did amt pab =
             newActiveBakers <- Trie.insert bid pad' (pab ^. activeBakers)
             return $ Right $ pab & activeBakers .~ newActiveBakers
 
+-- | Add a delegator to the persistent active bakers at a particular target.
+--  It is assumed that the delegator is not already delegated to this target.
+--  If the target is a baker, then the baker MUST be in the active bakers.
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/Bakers.hs:523]: Then it will error, as you mention. But it has already been checked earlier that baker is an active baker, so it will never happen.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/Accounts.hs [hunk 24]
```diff
@@ -420,6 +420,14 @@ updateAccountsAtIndex fupd ai a0@Accounts{..} =
         Nothing -> return (Nothing, a0)
         Just (res, act') -> return (Just res, a0{accountTable = act'})
 
+-- | Set the account at the given index. There must already be an account at the given index
+--  (otherwise this has no effect).
+setAccountAtIndex :: (SupportsPersistentAccount pv m) => AccountIndex -> PersistentAccount (AccountVersionFor pv) -> Accounts pv -> m (Accounts pv)
+setAccountAtIndex ai newAcct a0@Accounts{..} =
+    L.update (const (return ((), newAcct))) ai accountTable >>= \case
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/Accounts.hs]: I tried to make it a little bit less dense. The function passed to `L.update` needs to be a tuple, the first element being a return value, but in this case we just use want to set the account and not return anything.

Maybe have a look at `updateAccountsAtIndex` a few lines above to see how the return value in that case is used.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs (hunk 11 above)
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:1616]: Added comment that it is safe due to precondtion.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 25]
```diff
@@ -1747,220 +2035,496 @@ doConstrainBakerCommission pbs ai ranges = do
...
+    ) =>
+    PersistentBlockState (MPV m) ->
+    -- | Current block timestamp
+    Timestamp ->
+    AccountIndex ->
+    DelegatorUpdate ->
+    MTL.ExceptT DelegatorConfigureFailure m ([DelegationConfigureUpdateChange], PersistentBlockState (MPV m))
+newUpdateDelegator pbs blockTimestamp ai du@DelegatorUpdate{..} = do
+    bsp <- loadPBS pbs
+    -- Cannot fail: account must exist.
```
[the remark this answers is not in the document]
Comment (eb-concordium) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs]: Added explanation

**Review by eb-concordium — APPROVED** [review]
